packages/connectors/src/apps/discord/oauth.ts

```diff
@@ -0,0 +1,84 @@
+import { addSeconds } from "date-fns";
+
+import { expiresIn5Minutes } from "@rollout/connectors/utils/expires-in-5mins";
+import { defineRolloutOAuth2Service } from "@rollout/framework/rollout-oauth2";
+
+import { DiscordCredential } from "./auth";
+import { envSchema } from "./env";
+import { DiscordAPI } from "./lib/api";
+
+export const oauth = defineRolloutOAuth2Service<DiscordCredential>()({
+  env: envSchema,
+
+  pkce: true,
+
+  getAuthorizeUrl({ state, codeChallenge, redirectUri, env }) {
+    const params = new URLSearchParams({
+      state,
+      permissions: "8",
+      client_id: env.DISCORD_CLIENT_ID,
+      redirect_uri: redirectUri,
+      response_type: "code",
+      scope: "identify email guilds guilds.members.read bot",
+      code_challenge: codeChallenge!,
+    });
+
+    return `https://discord.com/api/oauth2/authorize?${params}`;
+  },
+
+  async getTokenFromCode({ code, codeVerifier, redirectUri, env }) {
+    const token = await fetch("https://discord.com/api/oauth2/token", {
+      method: "POST",
+      body: new URLSearchParams({
+        client_id: env.DISCORD_CLIENT_ID,
+        client_secret: env.DISCORD_CLIENT_SECRET,
+        grant_type: "authorization_code",
+        code,
+        redirect_uri: redirectUri,
+        code_verifier: codeVerifier!,
+      }),
+      headers: { "Content-Type": "application/x-www-form-urlencoded" },
+    }).then((res) => res.json());
+
+    const expireDate = addSeconds(new Date(), token.expires_in).toISOString();
+    const user = await DiscordAPI.getUser(token.access_token);
+
+    return {
+      credentialKey: user.id,
+      data: {
+        accessToken: token.access_token,
+        refreshToken: token.refresh_token,
+        expireDate,
+        guildId: token.guild.id,
```

Comment:
let's make the bot token part of the credential, that way we can then read it from credential data in lib.
```suggestion
        guildId: token.guild.id,
        botToken: env.DISCORD_BOT_TOKEN,
```